<?xml version="1.0"?>
<launch>
  <arg name="joy_dev" default="/dev/input/js0" />
  <arg name="nodelets_output" default="log" />
  <arg name="is_bag" default="false" />
  <arg name="plot_servo" default="false" />
  <arg name="obz_logfile" default="/home/autolab/ros16/outdoor/obz" />
  <arg name="use_alvar" default="false" />
  <arg name="use_obzerver" default="true" />
  <arg name="use_cftld" default="true" /> <!-- or fake it? -->
  <arg name="use_vicon" default="false" />
  <arg name="use_human" default="true" />
  <arg name="behavior_param_file" default="$(find bebop_hri)/config/behavior.yaml" />
  <arg name="drone_type" default="bebop2" /> <!-- available drone types: bebop1, bebop2 -->
  <arg name="advanced_close_range" default="true" />

  <group unless="$(arg is_bag)">
    <rosparam command="load" file="$(find bebop_hri)/config/bebop.yaml" />

    <include file="$(find bebop_tools)/launch/joy_teleop.launch">
      <arg name="joy_dev" value="$(arg joy_dev)" />
      <arg name="teleop_config" value="$(find bebop_hri)/config/log710.yaml" />
    </include>
    <node ns="bebop" pkg="bebop_driver" name="bebop_driver" type="bebop_driver_node" output="$(arg nodelets_output)">
      <param name="camera_info_url" value="package://bebop_driver/data/$(arg drone_type)_camera_calib.yaml"/>
      <param name="bebop_ip" value="192.168.42.1" />
      <rosparam command="load" file="$(find bebop_driver)/config/defaults.yaml" />
    </node>
  </group>

  <group if="$(arg use_alvar)">
    <arg name="marker_size" default="56.5" />
    <arg name="max_new_marker_error" default="0.08" />
    <arg name="max_track_error" default="0.2" />
    <arg name="cam_image_topic" default="bebop/image_raw" />
    <arg name="cam_info_topic" default="bebop/camera_info" />
    <arg name="output_frame" default="camera" />
    <arg name="alvar_max_freq" default="30.0" />

    <!--node name="republish" pkg="image_transport" type="republish" args="compressed in:=bebop/image_raw raw out:=bebop/image_raw" output="screen" /-->

    <node name="ar_track_alvar" pkg="ar_track_alvar" type="individualMarkersNoKinect" respawn="false" output="screen" args="$(arg marker_size) $(arg max_new_marker_error) $(arg max_track_error) $(arg cam_image_topic) $(arg cam_info_topic) $(arg output_frame) $(arg alvar_max_freq)">
      <remap from="roi" to="rqt_image_view/clicked_roi" />
    </node>
  </group>

  <include file="$(find bebop_hri)/launch/behavior.launch">
    <arg name="param_file" value="$(arg behavior_param_file)" />
  </include>

  <group if="$(arg use_obzerver)">
    <include file="$(find bebop_hri)/launch/obzerver.launch">
      <arg name="obz_logfile" value="$(arg obz_logfile)" />
      <arg name="start_paused" value="true" />
    </include>
  </group>

  <!-- this one is used by vservo node -->
  <include file="$(find bebop_vel_ctrl)/launch/bebop_vel_ctrl.launch">
    <arg name="param_file" value="$(find bebop_hri)/config/vel_ctrl_servo.yaml" />
    <arg name="output" value="log" />
    <arg name="input_topic" value="vservo/cmd_vel" />
    <arg name="output_topic" value="velctrl/cmd_vel" />
  </include>

  <!-- this one is used by the behavior node -->
  <include file="$(find bebop_vel_ctrl)/launch/bebop_vel_ctrl.launch">
    <arg name="name" value="abs_vel_ctrl" />
    <arg name="param_file" value="$(find bebop_hri)/config/vel_ctrl_abs.yaml" />
    <arg name="output" value="screen" />
    <arg name="input_topic" value="abs_vel_ctrl/setpoint/cmd_vel" />
    <arg name="output_topic" value="abs_vel_ctrl/cmd_vel" />
    <arg name="debug_topic" value="abs_vel_ctrl/debug" />
  </include>

  <include file="$(find bebop_vservo)/launch/bebop_vservo.launch">
    <arg name="param_file" value="$(find bebop_vservo)/config/bebop_vservo.yaml" />
    <arg name="output" value="screen" />
    <arg name="plot" value="$(arg plot_servo)" />
  </include>

  <node pkg="twist_mux" type="twist_mux" name="twist_mux" output="screen">
    <remap from="cmd_vel_out" to="/bebop/cmd_vel" />
    <rosparam file="$(find bebop_hri)/config/twist_mux.yaml" command="load"/>
    <rosparam file="$(find bebop_hri)/config/twist_mux_locks.yaml" command="load"/>
  </node>

  <group if="$(arg use_cftld)">
    <include file="$(find cftld_ros)/launch/cftld_node.launch">
      <!--arg name="output" value="$(arg nodelets_output)" /-->
      <arg name="output" value="screen" />
      <arg name="cftld_param_file" value="$(find cftld_ros)/config/default.yaml" />
      <arg name="cftld_config_file" value="$(find cftld_ros)/config/cftld.cfg" />
    </include>
  </group>

  <group unless="$(arg use_cftld)">
    <node name="fake_tracker" type="fake_tracker_   node" pkg="cftld_ros" output="log">
      <remap from="roi" to="rqt_image_view/clicked_roi" />
      <remap from="track" to="cftld/track" />
    </node>
  </group>

  <group if="$(arg use_human)">
    <include file="$(find autonomy_human)/launch/autonomy_human.launch">
      <arg name="param_file" value="$(find bebop_hri)/config/autonomy_human.yaml" />
      <arg name="input_topic" value="bebop/image_raw" />
      <arg name="output" value="screen" />
      <arg name="debug_mode" value="18" />
      <arg name="start_paused" value="true" />
    </include>
    <group if="$(arg advanced_close_range)">
      <node pkg="nodelet" type="nodelet" name="nodelet_manager" args="manager" output="screen"/>
      <!--node pkg="cftld_ros" type="cftld_ros_node" name="cftld_ros_right_hand_nodelet" output="screen" respawn="true">
        <remap from="image_raw" to="bebop/image_raw" />
        <remap from="init_roi" to="gesture/right_hand/cftld/init_roi" />
        <remap from="reset" to="gesture/right_hand/cftld/reset" />
        <remap from="track" to="gesture/right_hand/cftld/track" />
        <remap from="debug_image" to="gesture/right_hand/cftld/debug_image_raw" />
        <param name="tld_config_file" value="$(find cftld_ros)/config/cftld.cfg" />
        <rosparam command="load" file="$(find cftld_ros)/config/default.yaml" />
      </node>
      <node pkg="cftld_ros" type="cftld_ros_node" name="cftld_ros_left_hand_nodelet" output="screen" respawn="true">
        <remap from="image_raw" to="bebop/image_raw" />
        <remap from="init_roi" to="gesture/left_hand/cftld/init_roi" />
        <remap from="reset" to="gesture/left_hand/cftld/reset" />
        <remap from="track" to="gesture/left_hand/cftld/track" />
        <remap from="debug_image" to="gesture/left_hand/cftld/debug_image_raw" />
        <param name="tld_config_file" value="$(find cftld_ros)/config/cftld.cfg" />
        <rosparam command="load" file="$(find cftld_ros)/config/default.yaml" />
      </node-->
      <node pkg="nodelet" type="nodelet" name="yolo2" args="load yolo2/Yolo2Nodelet nodelet_manager" output="screen" respawn="true">
        <param name="confidence" value="0.5" />
        <param name="nms" value="0.4" />
      </node>
    </group>
  </group>

  <group if="$(arg use_vicon)">
    <include file="$(find bebop_hri)/launch/vicon.launch" />
  </group>
</launch>
